File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <iostream>
#include <vector>
using namespace std;

int n,s;
int res=0;
vector<int> stack;
int path[21]={0,};
int visited[21]={0,};
int sum=0;
void dfs(int lev, int level, int cri){
  if(lev==level){
    for(int i=0;i<lev;i++){
      sum+=path[i];
    }
    if(sum==s){
      
      res++;
    }
    sum=0;
    return;
  }

  for(int i=cri;i<n;i++){
    if(visited[i]==1) continue;
    path[lev]=stack[i];
    visited[i]=1;
    dfs(lev+1,level,i+1);
    path[lev]=0;
    visited[i]=0;
  }
}
int main(){
  cin>>n>>s;
  for(int i=0;i<n;i++){
    int num;
    cin>>num;
    stack.push_back(num);
  }
  for(int i=1;i<=n;i++){
    dfs(0,i,0);
  }
  cout<<res;
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
	public static int[] arr;
	public static int cnt = 0;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] tmp = br.readLine().split(" ");
		int n = Integer.parseInt(tmp[0]);
		int target = Integer.parseInt(tmp[1]);
		arr = new int[n];
		boolean[] visited = new boolean[n];
		int[] path = new int[n];
		String[] docs = br.readLine().split(" ");
		for(int i=0;i<n;i++) {
			arr[i] = Integer.parseInt(docs[i]);
		}
		for(int i=1;i<=n;i++) {
			dfs(0, 0, i, target, path, visited);
		}
		System.out.println(cnt);
	}
	
	public static void dfs(int cri, int lev, int level, int target, int[] path, boolean[] visited) {
		if(lev==level) {
			int sum = 0;
			for(int i=0;i<lev;i++) {
				sum += path[i];
			}
			if(sum == target) {
				cnt++;
			}
			return;
		}
		
		
		for(int i=cri;i<arr.length;i++) {
			if(!visited[i]) {
				visited[i] = true;
				path[lev] = arr[i];
				dfs(i+1, lev+1, level, target, path, visited);
				visited[i] = false;
			}
		}
	}
}
